### PR TITLE
task/vector_layers/drawer

### PR DESCRIPTION
**Pull request by teamagam5**
The drawer has a menu with items corresponding to layers on the map.

### CODE REVIEW

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/view/MainActivity.java [hunk 1]
```diff
@@ -215,52 +179,12 @@ private void initViewer() {
...
-        @Override
-        public void onDrawerClosed(View drawerView) {
-            sLogger.userInteraction("Drawer closed");
-        }
-
-        @Override
-        public void onDrawerStateChanged(int newState) {
-
-        }
-    }
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/view/MainActivity.java:265]: I'm looking at all this deleted code and my spirit goes up 👍

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+    public void displayHidden(VectorLayerPresentation vectorLayerPresentation) {
+        display(vectorLayerPresentation, false);
+    }
+
+    private void display(VectorLayerPresentation vectorLayerPresentation, boolean isVisible) {
+        if (!mStringToIntegerMapper.contains(vectorLayerPresentation.getId())) {
+            int intId = mStringToIntegerMapper.put(vectorLayerPresentation.getId());
+            ((MainActivityDrawer) mView).addToMenu(vectorLayerPresentation.getName(), intId);
+        }
+        ((MainActivityDrawer) mView).setChecked(
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: Casting can be avoided if you use the generic type of BaseViewModel for this class ->
extends BaseViewModel<MainAcitivtyDrawer>

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+    private DrawerLayout mDrawerLayout;
+
+    private StringToIntegerMapper mStringToIntegerMapper;
+
+    @Inject
+    public DrawerViewModel() {
+    }
+
+    public void start() {
+        mStringToIntegerMapper = new StringToIntegerMapper();
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: These two lines can be called at the constructor

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 4]
```diff
@@ -0,0 +1,128 @@
...
+        mDrawerStateListener = new DrawerStateListener();
+
+        mDisplayVectorLayersInteractor = mDisplayVectorLayersInteractorFactory.create(this);
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: This line can be also called at the constructor,
You can also replace inheritance with delegation (for the Displayer interface)

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 5]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    public void setDrawerMenuSelectedListenerTo(NavigationView navigationView) {
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: This method and the next one (setDrawerLayout) are expected to be called.
The class cannot function without them.
Therefore, ask for them at the object's creation.
As usual, either use a static factory that will make all the needed calls, or use AutoFactory for assisted injection

Comment (BarAgam) [reply app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: After that, you can change these methods to be private

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+        display(vectorLayerPresentation, true);
+    }
+
+    @Override
+    public void displayHidden(VectorLayerPresentation vectorLayerPresentation) {
+        display(vectorLayerPresentation, false);
+    }
+
+    private void display(VectorLayerPresentation vectorLayerPresentation, boolean isVisible) {
+        if (!mStringToIntegerMapper.contains(vectorLayerPresentation.getId())) {
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: It seems that this method does too much on its own.
I'd extract methods and refactor it like this:
1. addToMenuIfNecessary
2. getMenuItemId
3. setCheckedStatus

linear flow , simple english

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 7]
```diff
@@ -0,0 +1,128 @@
...
+        @Override
+        public void onDrawerSlide(View drawerView, float slideOffset) {
+
+        }
+
+        @Override
+        public void onDrawerOpened(View drawerView) {
+            sLogger.userInteraction("Drawer opened");
+
+            ((MainActivityDrawer) mView).setNavHeaderText(drawerView);
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: I know that code existed before, but should the navigation header text be set every time the drawer opens?
Couldn't it just be done on drawer's initialization?

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java [hunk 8]
```diff
@@ -0,0 +1,128 @@
...
+    }
+
+    private class DrawerItemSelectedListener
+            implements NavigationView.OnNavigationItemSelectedListener {
+        @Override
+        public boolean onNavigationItemSelected(@NonNull MenuItem item) {
+            sLogger.userInteraction("Drawer item " + item + " clicked");
+            switch (item.getGroupId()) {
+                case R.id.drawer_menu_layers_group:
+                    mSetVectorLayerVisibilityInteractorFactory.create(
```
Comment (BarAgam) [inline app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: extract method

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java [hunk 9]
```diff
@@ -6,6 +6,7 @@
 import com.teamagam.gimelgimel.domain.config.Constants;
 import com.teamagam.gimelgimel.domain.layers.ProcessIncomingVectorLayerInteractorFactory;
 import com.teamagam.gimelgimel.domain.location.respository.UsersLocationRepository;
+import com.teamagam.gimelgimel.domain.map.SetVectorLayerVisibilityInteractorFactory;
```
Comment (BarAgam) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java]: unused

File: domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java [hunk 10]
```diff
@@ -16,6 +17,7 @@
 import com.teamagam.gimelgimel.domain.messages.entity.MessageText;
 import com.teamagam.gimelgimel.domain.messages.entity.MessageUserLocation;
 import com.teamagam.gimelgimel.domain.messages.entity.MessageVectorLayer;
+import com.teamagam.gimelgimel.domain.messages.entity.contents.VectorLayer;
```
Comment (BarAgam) [inline domain/src/main/java/com/teamagam/gimelgimel/domain/messages/poller/PolledMessagesProcessor.java]: unused

**Review by BarAgam — APPROVED** [review]
Few simple changes.
Fix and merge.
You can also merge it to dev afterwards (rebase & merge)

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java (hunk 4 above)
Comment (teamagam5) [reply app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: I think the factory doesn't exist in the constructor (not yet injected?).
Anyway it doesn't work when moved to the constructor.

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java (hunk 2 above)
Comment (teamagam5) [reply app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: Cool!

File: app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java (hunk 7 above)
Comment (teamagam5) [reply app/src/main/java/com/teamagam/gimelgimel/app/mainActivity/viewmodel/DrawerViewModel.java]: The reason is that username wasn't immutable until now. This branch doesn't yet know about the new status of the username.

**Pull request merged**